File: src/features/typing/data/ts/word_break.ts
function wordBreak(s: string, wordDict: string[]): boolean {
    const words = new Set<string>(wordDict);
    const dp: boolean[] = new Array(s.length + 1).fill(false);
    dp[0] = true;
    for (let i = 1; i <= s.length; i++) {
        for (let j = 0; j < i; j++) {
            if (dp[j] && words.has(s.slice(j, i))) {
                dp[i] = true;
                break;
            }
        }
    }
    return dp[s.length];
}

File: src/features/typing/data/dart/word_break.dart
class Solution {
  bool wordBreak(String s, List<String> wordDict) {
    final words = wordDict.toSet();
    final dp = List<bool>.filled(s.length + 1, false);
    dp[0] = true;
    for (int i = 1; i <= s.length; i++) {
      for (int j = 0; j < i; j++) {
        if (dp[j] && words.contains(s.substring(j, i))) {
          dp[i] = true;
          break;
        }
      }
    }
    return dp[s.length];
  }
}
